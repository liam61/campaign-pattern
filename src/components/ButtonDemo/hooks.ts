import { useRequest } from 'ahooks'
import { isBoolean, isNil, isString } from 'lodash'

interface PermissionItem {
  code?: string | boolean
}

type BEPermissionMap = Partial<Record<typeof __LOCALE__, string[]>>

const fakePermissionMap: BEPermissionMap = {
  sg: ['CreateCampaign', 'EditCampaign'],
  cn: ['ViewCampaign'],
}

export const useDemoPermission = (item: PermissionItem): { loading: boolean; hasPerm: boolean } => {
  const { code } = item

  const { loading, data } = useRequest<BEPermissionMap | null, void[]>(
    async () => {
      await new Promise((resolve) => setTimeout(resolve, 500))

      return fakePermissionMap
    },
    {
      ready: isString(code), // when pass string, will load api
    }
  )

  if (isNil(code) || isBoolean(code)) {
    return {
      loading: false,
      hasPerm: !!code,
    }
  }

  return {
    loading,
    hasPerm: !!data?.[__LOCALE__]?.includes(code),
  }
}
