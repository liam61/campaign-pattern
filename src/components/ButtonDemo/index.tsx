import { ReactNode, useState } from 'react'

import { Button, ButtonProps } from 'antd'

import { useDemoPermission } from './hooks'

export * from './hooks'

type ButtonDemoProps = ButtonProps & {
  permission?: string | boolean
}

export function PermissionButtonDemo(props: ButtonDemoProps): ReactNode {
  const { permission, ...restProps } = props

  const { loading, hasPerm } = useDemoPermission({ code: permission })
  const [visible, setVisible] = useState(false)

  if (loading) return <div>loading...</div>
  if (!hasPerm) return null

  const handleClick = (): void => {
    setVisible(!visible)
  }

  return (
    <>
      <Button {...restProps} onClick={handleClick}>
        Button with Permission
      </Button>
      {!!visible && <div>Modal show</div>}
    </>
  )
}
