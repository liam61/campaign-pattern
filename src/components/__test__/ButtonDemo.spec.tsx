import { fireEvent, render, screen, waitFor } from '@testing-library/react'
import { renderHook } from '@testing-library/react-hooks'

import { PermissionButtonDemo, useDemoPermission } from '../ButtonDemo'

describe('useDemoPermission usages', () => {
  it('should directly return hasPerm false', () => {
    const { result } = renderHook(() => useDemoPermission({}))

    expect(result.current.hasPerm).toBe(false)
  })

  it('should directly return hasPerm false 2', () => {
    const { result } = renderHook(() => useDemoPermission({ code: false }))

    expect(result.current.hasPerm).toBe(false)
  })

  it('should directly return hasPerm true', () => {
    const { result } = renderHook(() => useDemoPermission({ code: true }))

    expect(result.current.hasPerm).toBe(true)
  })

  it('should first loading, then return true', async () => {
    const { result } = renderHook(() => useDemoPermission({ code: 'CreateMainCampaign' }))

    expect(result.current.loading).toBe(true)

    await waitFor(() => expect(result.current.loading).toBe(false))

    expect(result.current.hasPerm).toBe(true)
  })

  it('should first loading, then return false', async () => {
    const { result } = renderHook(() => useDemoPermission({ code: 'DeleteMainCampaign' }))

    expect(result.current.loading).toBe(true)

    await waitFor(() => expect(result.current.loading).toBe(false))

    // wrong region in this case
    expect(result.current.hasPerm).toBe(false)
  })
})

/**
 * see this doc for best practice
 * https://kentcdodds.com/blog/common-mistakes-with-react-testing-library
 *
 * see this about queries usage
 * https://testing-library.com/docs/queries/about
 */
describe('PermissionButtonDemo usages', () => {
  it('should directly render null', () => {
    const { container } = render(<PermissionButtonDemo permission={false} />)

    // will be a div container to wrap null
    expect(container).toMatchInlineSnapshot(`<div />`)
  })

  it('should directly render Button', () => {
    render(<PermissionButtonDemo permission={true} />)

    expect(screen.getByRole('button', { name: 'Button with Permission' })).toBeInTheDocument()
  })

  it('should first loading, then render Button', async () => {
    render(<PermissionButtonDemo permission="EditMainCampaign" />)

    expect(screen.getByText(/loading/i)).toBeInTheDocument()

    // use findBy*, it will wait until the component is found
    const buttonEl = await screen.findByRole('button', { name: 'Button with Permission' })
    // will auto generate and find snapshot
    expect(buttonEl).toMatchSnapshot()
  })

  it('should show modal text after clicking', async () => {
    render(<PermissionButtonDemo className="btn-flag" permission="EditMainCampaign" />)

    const buttonEl = await screen.findByRole('button', { name: 'Button with Permission' })

    expect(screen.queryByText(/Modal/i)).not.toBeInTheDocument()
    fireEvent.click(buttonEl)
    expect(screen.getByText(/Modal/i)).toBeInTheDocument()
  })
})
