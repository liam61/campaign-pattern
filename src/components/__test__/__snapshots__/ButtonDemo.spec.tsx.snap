// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PermissionButtonDemo usages > should first loading, then render Button 1`] = `
<button
  class="ant-btn ant-btn-default"
  type="button"
>
  <span>
    Button with Permission
  </span>
</button>
`;
